File: src/kubernetes/worker-pool.ts
import { all, ComponentResource, Output } from "@pulumi/pulumi";
import * as cloudinit from "@pulumi/cloudinit";
import * as equinix from "@equinix-labs/pulumi-equinix";
import * as fs from "fs";

import { Cluster } from "./cluster";
import { PREFIX } from "./meta";

type WorkerNode = equinix.metal.Device;

export interface Config {
  plan: equinix.metal.Plan;
  replicas: number;
}

export class WorkerPool extends ComponentResource {
  readonly cluster: Cluster;

  constructor(cluster: Cluster, name: string, config: Config) {
    super(
      `${PREFIX}:kubernetes:WorkerPool`,
      `${cluster.name}-${name}`,
      config,
      { parent: cluster }
    );

    this.cluster = cluster;

    for (let i = 1; i <= config.replicas; i++) {
      this.createWorkerPoolNode(name, cluster, config, i);
    }
  }

  createWorkerPoolNode(
    name: string,
    cluster: Cluster,
    config: Config,
    num: number
  ): WorkerNode {
    const device = new equinix.metal.Device(
      `${cluster.name}-${name}-${num}`,
      {
        hostname: `${cluster.name}-${name}-${num}`,
        metro: cluster.config.metro,
        billingCycle: equinix.metal.BillingCycle.Hourly,
        plan: config.plan,
        operatingSystem: equinix.metal.OperatingSystem.Ubuntu2204,
        projectId: cluster.config.project,
        customData: all([cluster.joinToken(), cluster.controlPlaneIp]).apply(
          ([joinToken, controlPlaneIp]) =>
            JSON.stringify({
              kubernetesVersion: cluster.config.kubernetesVersion,
              joinToken,
              controlPlaneIp,
            })
        ),
        userData: cloudConfig.then((c) => c.rendered),
      },
      {
        parent: this,
        dependsOn: cluster.controlPlane
          ? cluster.controlPlane.controlPlaneDevices.map((cp) => cp.device)
          : [],
      }
    );

    return device;
  }
}

const cloudConfig = cloudinit.getConfig({
  gzip: false,
  base64Encode: false,
  parts: [
    {
      contentType: "text/x-shellscript",
      content: fs.readFileSync(
        "../cloud-init/scripts/pre-install.sh",
        "utf8"
      ),
    },
    {
      contentType: "text/x-shellscript",
      content: fs.readFileSync(
        "../cloud-init/scripts/download-metadata.sh",
        "utf8"
      ),
    },
    {
      contentType: "text/x-shellscript",
      content: fs.readFileSync(
        "../cloud-init/scripts/containerd-prerequisites.sh",
        "utf8"
      ),
    },
    {
      contentType: "text/x-shellscript",
      content: fs.readFileSync(
        "../cloud-init/scripts/kubernetes-prerequisites.sh",
        "utf8"
      ),
    },
    {
      contentType: "text/x-shellscript",
      content: fs.readFileSync(
        "../cloud-init/scripts/kubelet-config.sh",
        "utf8"
      ),
    },
    {
      contentType: "text/x-shellscript",
      content: fs.readFileSync(
        "../cloud-init/scripts/kubernetes-kubeadm-packages.sh",
        "utf8"
      ),
    },
    {
      contentType: "text/x-shellscript",
      content: fs.readFileSync(
        "../cloud-init/scripts/kubernetes-kubeadm-worker-config.sh",
        "utf8"
      ),
    },
    {
      contentType: "text/x-shellscript",
      content: fs.readFileSync(
        "../cloud-init/scripts/kubernetes-kubeadm-worker-join.sh",
        "utf8"
      ),
    },
    {
      contentType: "text/x-shellscript",
      content: fs.readFileSync(
        "../cloud-init/scripts/net-deny-metadata.sh",
        "utf8"
      ),
    },
    {
      contentType: "text/x-shellscript",
      content: fs.readFileSync(
        "../cloud-init/scripts/post-install.sh",
        "utf8"
      ),
    },
  ],
});

File: python/src/kubernetes/worker_pool.py
import helpers
import pulumi_cloudinit as cloudinit
import pulumi_equinix as equinix
from pulumi import ComponentResource, Output, ResourceOptions, input_type

from .meta import PREFIX


class WorkerNode:
    def __init__(self, device):
        self.device = device


@input_type
class Config:
    def __init__(self, name_suffix, plan, replicas):
        self.name_suffix = name_suffix
        self.plan = plan
        self.replicas = replicas


class WorkerPool(ComponentResource):
    def __init__(self, cluster, config: Config):
        super().__init__(
            f"{PREFIX}:kubernetes:WorkerPool",
            f"{cluster.name}-{config.name_suffix}",
            config.__dict__,
            ResourceOptions(parent=cluster),
        )

        self.cluster = cluster

        for i in range(1, config.replicas + 1):
            self.__create_worker_pool_node(config.name_suffix, cluster, config, i)

    def __create_worker_pool_node(self, name: str, cluster, config: Config, num: int):
        device = equinix.metal.Device(
            f"{cluster.name}-{name}-{num}",
            hostname=f"{cluster.name}-{name}-{num}",
            metro=cluster.config.metro,
            billing_cycle=equinix.metal.BillingCycle.HOURLY,
            plan=config.plan,
            operating_system=equinix.metal.OperatingSystem.UBUNTU2204,
            project_id=cluster.config.project,
            custom_data=Output.all(cluster.join_token, cluster.control_plane_ip).apply(
                lambda values: Output.json_dumps(
                    {
                        "kubernetesVersion": cluster.config.kubernetes_version,
                        "joinToken": values[0],
                        "controlPlaneIp": values[1],
                    },
                    separators=(",", ":"),
                )
            ),
            user_data=cloud_config.rendered,
            opts=ResourceOptions(
                parent=self,
                depends_on=list(
                    map(
                        lambda cp: cp.device,
                        cluster.control_plane.control_plane_devices,
                    )
                ),
            )
            if cluster.control_plane
            else None,
        )

        return WorkerNode(device)


cloud_config = cloudinit.get_config(
    gzip=False,
    base64_encode=False,
    parts=[
        cloudinit.ConfigPartArgs(
            content_type="text/x-shellscript",
            content=helpers.get_file_content(
                f"{helpers.get_project_root()}/cloud-init/scripts/pre-install.sh"
            ),
        ),
        cloudinit.ConfigPartArgs(
            content_type="text/x-shellscript",
            content=helpers.get_file_content(
                f"{helpers.get_project_root()}/cloud-init/scripts/download-metadata.sh"
            ),
        ),
        cloudinit.ConfigPartArgs(
            content_type="text/x-shellscript",
            content=helpers.get_file_content(
                f"{helpers.get_project_root()}/cloud-init/scripts/containerd-prerequisites.sh"
            ),
        ),
        cloudinit.ConfigPartArgs(
            content_type="text/x-shellscript",
            content=helpers.get_file_content(
                f"{helpers.get_project_root()}/cloud-init/scripts/kubernetes-prerequisites.sh"
            ),
        ),
        cloudinit.ConfigPartArgs(
            content_type="text/x-shellscript",
            content=helpers.get_file_content(
                f"{helpers.get_project_root()}/cloud-init/scripts/kubernetes-kubeadm-packages.sh"
            ),
        ),
        cloudinit.ConfigPartArgs(
            content_type="text/x-shellscript",
            content=helpers.get_file_content(
                f"{helpers.get_project_root()}/cloud-init/scripts/kubernetes-kubeadm-worker-config.sh"
            ),
        ),
        cloudinit.ConfigPartArgs(
            content_type="text/x-shellscript",
            content=helpers.get_file_content(
                f"{helpers.get_project_root()}/cloud-init/scripts/kubernetes-kubeadm-worker-join.sh"
            ),
        ),
        cloudinit.ConfigPartArgs(
            content_type="text/x-shellscript",
            content=helpers.get_file_content(
                f"{helpers.get_project_root()}/cloud-init/scripts/net-deny-metadata.sh"
            ),
        ),
        cloudinit.ConfigPartArgs(
            content_type="text/x-shellscript",
            content=helpers.get_file_content(
                f"{helpers.get_project_root()}/cloud-init/scripts/post-install.sh"
            ),
        ),
    ],
)
